msal/src/test/java/com/microsoft/identity/client/e2e/tests/mocked/GetDeviceCodeTest.java

```diff
@@ -0,0 +1,155 @@
+
+        Assert.assertNotNull(authorizationResponse.getDeviceCode());
+        Assert.assertNotNull(authorizationResponse.getUserCode());
+        Assert.assertNotNull(authorizationResponse.getMessage());
+        Assert.assertNotNull(authorizationResponse.getInterval());
+        Assert.assertNotNull(authorizationResponse.getExpiresIn());
+        Assert.assertNotNull(authorizationResponse.getVerificationUri());
+
+        Assert.assertNull(authorizationResult.getAuthorizationErrorResponse());
+    }
+
+    @Test
+    public void testGetDeviceCodeFailureNoClientId() {
+        final MicrosoftStsAuthorizationRequest authorizationRequest = builder.setClientId(null).build();
+        final AuthorizationResult authorizationResult = runGetDeviceCodeThread(authorizationRequest, urlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals("invalid_request", authorizationErrorResponse.getError());
+    }
+
+    @Test
+    public void testGetDeviceCodeFailureNoScope() {
+        final MicrosoftStsAuthorizationRequest authorizationRequest = builder.setScope(null).build();
+        final AuthorizationResult authorizationResult = runGetDeviceCodeThread(authorizationRequest, urlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals("invalid_request", authorizationErrorResponse.getError());
+    }
+
+    @Test
+    public void testGetDeviceCodeFailureBadScope() {
+        final MicrosoftStsAuthorizationRequest authorizationRequest = builder.setScope("/").build();
+        final AuthorizationResult authorizationResult = runGetDeviceCodeThread(authorizationRequest, urlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals("invalid_scope", authorizationErrorResponse.getError());
+    }
+
+    /**
+     * Helper function to run getDeviceCode(). Catches exception
+     * @param authorizationRequest request to send to getDeviceCode()
+     * @param urlBody url to send to getDeviceCode()
```

Comment:
Where are we using ShadowDeviceCodeFlowCommandTokenError?

Also can we have test cases for each expected error from the token endpoint?
 authorization_declined, bad_verification_code, expired_token

Reply:
`ShadowDeviceCodeFlowCommandTokenError` is used in `DeviceCodeFlowAPITest`. I was thinking of moving my getDeviceCode tests into that file and then adding the token tests into there as well.